Enable Rune Fencer Category 2 Merits, Implement Vivacious Pulse & Swordplay, Fix level for Tactical Parry III on RUN

WinterSolstice8 opened the pull request:
Please see attached for this branches progress once merged. 
[RUN progress.pdf](https://github.com/LandSandBoat/server/files/8149873/RUN.progress.pdf)

I got a lot of work done and decided it was time to merge back in since it was getting a little beefy...

This PR implements
* Vivacious Pulse (complete, minus dynamis divergence weapon aug)
* Sleight of Sword (complete, minus buff wiping on taking "severe" dmage)
* Swordplay (Complete)
* Enable Rune Fencer Category 2 merits
* Fix level for Tactical Parry III for RUN
* Fix Inquartata gifts & add Inquartata to Status.lua as there are gear mods that use it

Edit: looks like I accidentally included a file in the last commit that modifies scripts/commands/getstats.lua to include Subtle Blow and Store TP in the display... though I suspect that won't be a problem, other than maybe amending the commit message.

**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

WinterSolstice8 committed:
Enable Rune Fencer category 2 merits

WinterSolstice8 committed:
Add Inquartata to status.lua, fix job_point_gifts.sql to point to correct effect

WinterSolstice8 committed:
Fix Tactical Parry III level for RUN
* source: https://www.bg-wiki.com/ffxi/Tactical_Parry

WinterSolstice8 committed:
Implement Swordplay and Vivacious Pulse
* Vivacious pulse is missing Morgelai etc aug bonus due to the aug value not being known at this time.

WinterSolstice8 committed:
Remove whitespace addition

Xaver-DaRed commented on scripts/commands/getstats.lua:
```diff
@@ -49,6 +49,8 @@ function onTrigger(player)
     player:PrintToPlayer(string.format("Magic Defense bonus: %i ", target:getMod(xi.mod.MDEF)), xi.msg.channel.SYSTEM_3)
     player:PrintToPlayer(string.format("Magic Accuracy bonus: %i ", target:getMod(xi.mod.MACC)), xi.msg.channel.SYSTEM_3)
+    player:PrintToPlayer(string.format("Total Subtle Blow: %i ", target:getMod(xi.mod.SUBTLE_BLOW)), xi.msg.channel.SYSTEM_3)
+    player:PrintToPlayer(string.format("Total Store TP: %i ", target:getMod(xi.mod.STORETP)), xi.msg.channel.SYSTEM_3)
```
I don't personaly care for this being here, but I would still suggest to at least comment them out.

Xaver-DaRed commented on scripts/globals/abilities/swordplay.lua:
```diff
@@ -0,0 +1,22 @@
+-- Recast Time: 5:00
+-- Duration: 2:00
+-----------------------------------
+require("scripts/settings/main")
```
Uneeded require here.

Xaver-DaRed commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,166 @@ local function enforceRuneCounts(target)
+
+
+
+--source https://www.bg-wiki.com/ffxi/Vivacious_Pulse
```
Lot's of empty space between functions.

Xaver-DaRed commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,166 @@ local function enforceRuneCounts(target)
+local function calculateVivaciousPulseHealing(target)
+
+    local divineMagicSkillLevel = target:getSkillLevel(xi.skill.DIVINE_MAGIC)
+    local HPHealAmount = 10 + math.floor(divineMagicSkillLevel / 2 * (100 + target:getJobPointLevel(xi.jp.VIVACIOUS_PULSE_EFFECT)) / 100) -- Bonus of 1-20%  from Vivacious pulse job points.
```
Just I random thought, feel free to ignore it.
I think divineMagicSkillLevel / 2 should be floored if its gonna get multiplied right after. SE usually floors after each step.
I have no proof of this though.

WinterSolstice8 replied:
I went and checked retail, and I got a second opinion confirming it, see the following post and the post after
https://www.ffxiah.com/forum/topic/37086/endeavoring-to-awaken-a-guide-to-rune-fencer/180/#3615355

Xaver-DaRed commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,166 @@ local function enforceRuneCounts(target)
+            or type == xi.effect.CURSE_II or type == xi.effect.DOOM or type == xi.effect.VIRUS
+            or type == xi.effect.PLAGUE or type == xi.effect.PETRIFICATION then
+             debuffs[debuffCount+1] = type
+             debuffCount = debuffCount + 1
```
The elseif isn't properly aligned. (Might be just in my end though, please confirm, just in case.)
Also, we normaly leave the "and" / "or" at the end of the line, like...
```
    elseif
        type == xi.effect.POISON or type == xi.effect.PARALYSIS or type == xi.effect.BLINDNESS or
        type == xi.effect.SILENCE or type == xi.effect.MUTE or type == xi.effect.CURSE_I or
        type == xi.effect.CURSE_II or type == xi.effect.DOOM or type == xi.effect.VIRUS or
        type == xi.effect.PLAGUE or type == xi.effect.PETRIFICATION
    then
```

WinterSolstice8 replied:
Thanks for the tip on the styling there. I couldn't decide how I wanted it but now that I see that it makes sense.

Xaver-DaRed commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,166 @@ local function enforceRuneCounts(target)
+
+
+    -- see https://www.bg-wiki.com/ffxi/Sleight_of_Sword for levels
+    local meritBonus =  player:getMerit(xi.merit.MERIT_SLEIGHT_OF_SWORD)
```
Alot of empty space both here and between the start of this function and the end of previous

TeoTwawki commented:
> Please see attached for this branches progress once merged.
[RUN progress.pdf](https://github.com/LandSandBoat/server/files/8149873/RUN.progress.pdf)

You could use an issue for tracking. Github's markdown parsing makes it easy to make checklists and tables and such

Xaver-DaRed commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,166 @@ local function enforceRuneCounts(target)
+    end
+
+    if power > 0 then -- add aug bonus if appropriate gear is equipped. Note: ilvl 109+ "relic" or "AF2" gear always has the augment, so no need to check extdata. RUN does not have AF/AF2/AF3 gear below i109.
+            augBonus = (meritBonus / 5) * 2
```
This line isn't properly aligned.

Xaver-DaRed reviewed (dismissed):
Mostly styling. Logic looks proper at first glance. Great work.

WinterSolstice8 commented:
> > Please see attached for this branches progress once merged.
> > [RUN progress.pdf](https://github.com/LandSandBoat/server/files/8149873/RUN.progress.pdf)
> 
> You could use an issue for tracking. Github's markdown parsing makes it easy to make checklists and tables and such

Good idea. That way it's not just sitting on my local computer. Plus fancier when I can directly reference the issue number.

WinterSolstice8 committed:
Whitespace & formatting cleanup
* Comment unneeded additions to scripts/commands/getstats.lua

WinterSolstice8 commented:
Transferred that PDF to here.
#1340

zach2good commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,163 @@ local function enforceRuneCounts(target)
+        local tickPower = 3 -- Tick power appears to be 3/tick, not 6/tick if RUN main and 3/tick if RUN sub; source : https://www.ffxiah.com/forum/topic/37086/endeavoring-to-awaken-a-guide-to-rune-fencer/180/#3615377
+        local handsSlotID = target:getEquipID(xi.slot.HANDS)
+
+        -- add starting tick bonuses if appropriate gear is equipped
+        if handsSlotID      == 27018 then
+            power = 3 * tickPower
+        elseif handsSlotID  == 27019 then
+            power = 5 * tickPower
+        elseif handsSlotID  == 23218 then
+            power = 7 * tickPower
+        elseif handsSlotID  == 23553 then
+            power = 9 * tickPower
+        end
+    end
```
These will need to be extracted out to an item mod, if you let us know what you need here we can push those changes straight into main so you don't get any conflicts while this branch is ongoing

WinterSolstice8 replied:
![image](https://user-images.githubusercontent.com/60417494/155935541-35389d37-06dd-4447-a8f0-0f5b95133f61.png)
"Swordplay" item mod makes sense. I don't think anything else but this particular job specific gear boosts this currently though. These item IDs are all from Futhark Mitons/Futhark Mitons +1/Futhark Mitons +2/Futhark Mitons +3. We'd also need to get the extdata for Sleight of Sword if we replace that with an item mod, since we are no longer "guaranteed" to have the item equipped that will always have said augment. That, or the ID check stays but only for the Sleight of Sword augment bonus.

I have Futhark Mitons +3 in retail so i'll be able to dump the packet/data for the augment if I get instructions on how to do so. I assume NQ and all HQs all share the same augment...

TeoTwawki replied:
We -always- want to avoid checking item IDs when possible. And there's always the Chance SE adds something new that would use the mod.

> We'd also need to get the extdata for Sleight of Sword if we replace that with an item mod,

That's not that bad. The part that may be a problem is if its in the segment of augment IDs we handle incorrectly atm (some don't display right, but its usually the type that use a the other items typing with the sockets or slotting). That said, I would bet on its ID being nearby the other job-ability enchanting augments already in our table. Once we have the ID its just a matter of filling in the modifier info for it in the aforementioned table.

TeoTwawki replied:
I'll see if I can find a cap or even dig through the values for it later.

TeoTwawki replied:
Turns out we already have the ID
```sql
INSERT INTO `augments` VALUES (1457,0,0,0,0,0); -- Enhances "Sleight of Sword" effect
```
just needs modifier hooked up where those zeroes are, in `augments.sql`

WinterSolstice8 replied:
yup, client agrees. I added that aug to an !additem call.
![image](https://user-images.githubusercontent.com/60417494/155938914-b02d2d63-d6c4-4e59-8d2b-1dd9b9039813.png)

I will need mods for the following:

- Enhances "Battuta" effect
- Enhances "Elemental Sforzo" effect 
- Enhances "Sleight of Sword" effect
- Enhances "Inspiration" effect 
- Enhances "Elemental Sforzo" effect 
- Swordplay
- Liement
- "Enhances "Valiance" and "Vallation" effects" (they are literally the same buff, but one is AoE. Only one mod is needed.)
- Pflug
- "Liement" effect extends to an area
- "Vivacious Pulse" potency (looks like the augment is 1310 according to windower resources, but I can't get Morgelai to display it. Probably some nonsense with the Path A/B/C extdata that I know very little about)
EDIT:
Also need
- Augments "Vivacious Pulse"

TeoTwawki replied:
> "Vivacious Pulse" potency (looks like the augment is 1310 according to windower resources, but I can't get Morgelai to display it. Probably some nonsense with the Path A/B/C extdata that I know very little about)

Windower's resource extractor pulls the dat file index, while our table uses the order the packet does - the 2 lists don't agree and we currently lack a way to match them up. Take a look in the augments.sql file and use the `!additem` GM command to try nearby IDs starting from where "Slight of Sword" was - its probably nearby.

claywar commented on scripts/commands/getstats.lua:
```diff
@@ -49,6 +49,8 @@ function onTrigger(player)
     player:PrintToPlayer(string.format("Magic Attack bonus: %i ", target:getMod(xi.mod.MATT)), xi.msg.channel.SYSTEM_3)
     player:PrintToPlayer(string.format("Magic Defense bonus: %i ", target:getMod(xi.mod.MDEF)), xi.msg.channel.SYSTEM_3)
     player:PrintToPlayer(string.format("Magic Accuracy bonus: %i ", target:getMod(xi.mod.MACC)), xi.msg.channel.SYSTEM_3)
+    -- player:PrintToPlayer(string.format("Total Subtle Blow: %i ", target:getMod(xi.mod.SUBTLE_BLOW)), xi.msg.channel.SYSTEM_3)
```
I have no problem with these being uncommented, seems like a good addition to the gm command.

claywar commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,163 @@ local function enforceRuneCounts(target)
+        return 0
+    end
+
+    local runeStatMap = {
```
Just a minor styling preference, but I would suggest something more left aligned with allman braces:
```
local runeStatMap =
{
    [xi.effect.IGNIS]  = xi.mod.STR,
...
}

claywar commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,163 @@ local function enforceRuneCounts(target)
+                            [xi.effect.LUX]    = xi.mod.CHR,
+                        }
+
+    local stat = runeStatMap[effect:getType()]
```
If the below is possible to be nil, this will throw an exception and break out of the function if getType does not exist in the above table.

WinterSolstice8 replied:
I adjusted this logic so that nil check is no longer necessary. Take a look!

claywar commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,163 @@ local function enforceRuneCounts(target)
+        if type >= xi.effect.IGNIS and type <= xi.effect.TENEBRAE then
+            HPHealAmount = HPHealAmount + getRuneHealAmount(effect, target)
+        elseif
+            type == xi.effect.POISON or type == xi.effect.PARALYSIS or
```
Would suggest tabling up these items and looping over them as well instead of a very long conditional.

WinterSolstice8 replied:
I adjusted it to a table lookup, like the rune element mapping. tell me what you think

claywar commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,163 @@ local function enforceRuneCounts(target)
+        target:addMP(MPHealAmount) -- augment bonusPct does not apply here according to testing.
+    end
+
+    if debuffCount > 0 and (target:getEquipID(xi.slot.HEAD) == 26782 or target:getEquipID(xi.slot.HEAD) == 26783) then -- add random removal of Poison, Paralyze, Blind, Silence, Mute, Curse, Bane, Doom, Virus, Plague, Petrification via AF3 head (source: https://www.bg-wiki.com/ffxi/Erilaz_Galea)
```
Magic numbers for Item ID should probably be added to the enum in scripts/globals/items.lua

TeoTwawki replied:
mods!

claywar commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,163 @@ local function enforceRuneCounts(target)
+    end
+
+    HPHealAmount = HPHealAmount * bonusPct
+    if (target:getHP() + HPHealAmount > target:getMaxHP()) then
```
Minor styling, no need for outer parens here.

claywar commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,163 @@ local function enforceRuneCounts(target)
+        local handsSlotID = target:getEquipID(xi.slot.HANDS)
+
+        -- add starting tick bonuses if appropriate gear is equipped
+        if handsSlotID      == 27018 then
```
More things that could go into items.lua

TeoTwawki replied:
> More things that could go into items.lua

nay, mods mods mods

WinterSolstice8 replied:
I'm punting magic number changes until we get the appropriate item mods into the main branch. It will get fixed in a future PR, and I will note it in the progress chart. They will exist here as reminders for _what_ items need to have their sql entries updated. Don't need to duplicate work especially when we know the better way is on the horizon.

claywar reviewed (commented):
Looking like a great continuation of the previous PRs.  Left a few comments for some additional items. 👍

WinterSolstice8 committed:
Cleanup rune_fencer.lua
* Adjust logic of Vivacious Pulse effect a little

Xaver-DaRed reviewed (approved):
If more work is OTW in future PRs, I don't see a problem leaving it as it is here.

zach2good commented:
I'll do the mod number additions tonight when I finish work 👍

zach2good commented:
Added this PR: https://github.com/LandSandBoat/server/pull/1353

It isn't too bad to add new Mods, you just need to look at the bottom of the C++ list, find the next free number, add them, update the comment and copy things across to the Lua side 👍

zach2good commented:
Rebased `feature/runefencer` with the new changes in `main` 👍

WinterSolstice8 committed:
Change hardcoded gear checks to xi.mod checks
* Actually apply healing in Vivacious Pulse and not just the log message.

WinterSolstice8 commented:
Not entirely sure why it thinks that many of the commits duplicated...

zach2good commented on scripts/globals/job_utils/rune_fencer.lua:
```diff
@@ -50,8 +50,167 @@ local function enforceRuneCounts(target)
+
+xi.job_utils.rune_fencer.useVivaciousPulse = function(player, target, ability, effect)
+    return calculateVivaciousPulseHealing(player, target)
+end
```
Newline at end of file please

zach2good reviewed (approved):
LGTM, just some style nits, and the style checker is mad at you:
```
    scripts/globals/status.lua:1219:5: (W314) value assigned to field INQUARTATA is overwritten on line 1578 before use
```

If don't want to fight your commits, we can just squash-on-merge, but I'm sure you probably want the credit for multiple commits :P

WinterSolstice8 commented:
Legitimately have no idea where that redefinition of inquartata came from... I just rebased with upstream/base and never touched that file to my memory. will dig into it

WinterSolstice8 commented:
> Legitimately have no idea where that redefinition of inquartata came from... I just rebased with upstream/base and never touched that file to my memory. will dig into it
Ah... I see. I added it myself because it had already existed in the c++ code but not status. Then came along #1353 which happened to merge without errors back into the same exact location. mystery solved. I will just have to remove it on my end.

WinterSolstice8 committed:
Remove duplicate status in status.lua, update RUN status descriptions.

WinterSolstice8 committed:
Whitespace cleanup

zach2good commented:
Content-wise, this looks good and it's ready to go in. But, we can't take all of those duplicate commits. I also don't want to cheat you out of a bunch of commits by squashing everything down into a single commit. Could you rebase and fix your commits?

WinterSolstice8 commented:
Agreed, I think it's a bit of a mess too. I will figure it out after work/this weekend.

WinterSolstice8 commented:
I rebased off upstream/feature/rune-fencer, then cherry picked the commits that actually were meant to represent the final product.
